dpc/src/record/encoded.rs

```diff
@@ -16,45 +16,90 @@
 
 use crate::{DPCError, EncodedRecordScheme, Parameters, Payload, Record, RecordScheme};
 use snarkvm_algorithms::{
-    encoding::Elligator2,
     traits::{CommitmentScheme, CRH},
+    EncodingError,
 };
-use snarkvm_curves::traits::{AffineCurve, MontgomeryParameters, ProjectiveCurve, TwistedEdwardsParameters};
-use snarkvm_fields::PrimeField;
-use snarkvm_utilities::{
-    from_bits_le_to_bytes_le,
-    from_bytes_le_to_bits_le,
-    to_bytes_le,
-    BigInteger,
-    FromBytes,
-    ToBytes,
-};
+use snarkvm_fields::{FieldParameters, PrimeField};
+use snarkvm_utilities::{io::Cursor, to_bytes_le, FromBits, FromBytes, Read, ToBits, ToBytes};
+
+/// Encode a base field element bytes to a group representation.
+pub fn encode_to_field<F: PrimeField>(x_bytes: &[u8]) -> Result<Vec<F>, DPCError> {
+    let capacity = <F::Parameters as FieldParameters>::CAPACITY as usize;
+
+    let mut bits = Vec::<bool>::with_capacity(x_bytes.len() * 8 + 1);
+    for byte in x_bytes.iter() {
+        let mut byte = byte.clone();
+        for _ in 0..8 {
+            bits.push(byte & 1 == 1);
+            byte >>= 1;
+        }
+    }
+    bits.push(true);
+    // The last bit indicates the end of the actual data, which is used in decoding to
+    // make sure that the length is correct.
 
-use itertools::Itertools;
-use std::marker::PhantomData;
-
-/// Encode a base field element bytes to a group representation
-pub fn encode_to_group<P: MontgomeryParameters + TwistedEdwardsParameters, G: ProjectiveCurve>(
-    x_bytes: &[u8],
-) -> Result<(<G as ProjectiveCurve>::Affine, bool), DPCError> {
-    // TODO (howardwu): Remove this hardcoded value and use BaseField's size in bits to pad length.
-    let mut bytes = x_bytes.to_vec();
-    while bytes.len() < 32 {
-        bytes.push(0)
+    let mut res = Vec::<F>::with_capacity((bits.len() + capacity - 1) / capacity);
+    for chunk in bits.chunks(capacity) {
+        res.push(F::from_repr(F::BigInteger::from_bits_le(chunk)).unwrap());
     }
 
-    let input = P::BaseField::read_le(&bytes[..])?;
-    let (output, fq_high) = Elligator2::<P, G>::encode(&input)?;
-    Ok((output, fq_high))
+    Ok(res)
 }
 
-/// Decode a group into the byte representation of a base field element
-pub fn decode_from_group<P: MontgomeryParameters + TwistedEdwardsParameters, G: ProjectiveCurve>(
-    affine: <G as ProjectiveCurve>::Affine,
-    fq_high: bool,
-) -> Result<Vec<u8>, DPCError> {
-    let output = Elligator2::<P, G>::decode(&affine, fq_high)?;
-    Ok(to_bytes_le![output]?)
+/// Decode a group into the byte representation of a base field element.
+pub fn decode_from_field<F: PrimeField>(field_elements: &[F]) -> Result<Vec<u8>, DPCError> {
+    if field_elements.len() == 0 {
```

Comment:
```suggestion
    if field_elements.is_empty() {
```
otherwise `clippy` might complain about it